crates/factor-outbound-http/src/wasi.rs

```diff
@@ -61,10 +70,261 @@ impl<'a> WasiHttpView for WasiHttpImplInner<'a> {
     ) -> wasmtime_wasi_http::HttpResult<wasmtime_wasi_http::types::HostFutureIncomingResponse> {
-        // TODO: port implementation from spin-trigger-http
-        Err(ErrorCode::HttpRequestDenied.into())
+        if let Some(interceptor) = &self.state.request_interceptor {
+            let intercepted = Intercepted {
+                request: &mut request,
+                config: &mut config,
+            };
+            if let Some(res) = interceptor.intercept(intercepted) {
+                return res;
+            }
+        }
+
+        let host = request.uri().host().unwrap_or_default();
+        let tls_client_config = self
+            .state
+            .component_tls_configs
+            .get_client_config(host)
+            .clone();
+
+        Ok(HostFutureIncomingResponse::Pending(
+            wasmtime_wasi::runtime::spawn(
+                send_request_impl(
+                    request,
+                    config,
+                    self.state.allowed_hosts.clone(),
+                    tls_client_config,
+                )
+                .in_current_span(),
+            ),
+        ))
     }
 }
+
+async fn send_request_impl(
+    mut request: Request<wasmtime_wasi_http::body::HyperOutgoingBody>,
+    mut config: wasmtime_wasi_http::types::OutgoingRequestConfig,
+    allowed_hosts: OutboundAllowedHosts,
+    tls_client_config: Arc<ClientConfig>,
+) -> anyhow::Result<Result<IncomingResponse, ErrorCode>> {
+    let allowed_hosts = allowed_hosts.resolve().await?;
+
+    let is_relative_url = request.uri().authority().is_none();
+    if is_relative_url {
+        if !allowed_hosts.allows_relative_url(&["http", "https"]) {
+            return handle_not_allowed(request.uri(), true);
+        }
+
+        let origin = request
+            .extensions()
+            .get::<SelfRequestOrigin>()
+            .cloned()
+            .context("cannot send relative outbound request; no 'origin' set by host")?;
+
+        config.use_tls = origin.use_tls();
+
+        request.headers_mut().insert(HOST, origin.host_header());
+
+        let path_and_query = request.uri().path_and_query().cloned();
+        *request.uri_mut() = origin.into_uri(path_and_query);
+    } else {
+        let outbound_url = OutboundUrl::parse(request.uri().to_string(), "https")
+            .map_err(|_| ErrorCode::HttpRequestUriInvalid)?;
+        if !allowed_hosts.allows(&outbound_url) {
+            return handle_not_allowed(request.uri(), false);
+        }
+    }
+
+    if let Some(authority) = request.uri().authority() {
+        let current_span = tracing::Span::current();
+        current_span.record("server.address", authority.host());
+        if let Some(port) = authority.port() {
+            current_span.record("server.port", port.as_u16());
+        }
+    }
+
+    Ok(send_request_handler(request, config, tls_client_config).await)
+}
+
+// TODO(factors): Move to some callback on spin-factor-outbound-networking (?)
```

Comment:
I'd like to figure this out sooner rather than later since we're requiring a fairly Spin CLI specific logic (e.g., the use of `terminal::warn`) on every user of this factor. It would be nice to ensure we can handle this outside of the factor.